File: problems/chapter06/miiitomi/007.cpp
// AtCoder ARC101   D - Median of Medians
// https://atcoder.jp/contests/arc101/tasks/arc101_b

#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> a;

template <typename T>
struct BIT {
    int n;
    vector<T> data;

    BIT(int n_) {
        n = n_ + 1;
        data.assign(n, 0);
    }

    void add(int x, T a) {
        for (int i = x + 1; i < n; i += (i & -i)) {
            data[i] += a;
        }
    }

    T sum(int x) {
        T s = 0;
        for (int i = x + 1; i > 0; i -= (i & -i)) {
            s += data[i];
        }
        return s;
    }
};

bool f(int x) {
    vector<int> b(N);
    for (int i = 0; i < N; i++) {
        if (a[i] >= x) b[i] = 1;
        else b[i] = -1;
    }

    vector<int> s(N+1, 0);
    for (int i = 0; i < N; i++) {
        s[i+1] = s[i] + b[i];
    }

    int min_s = *min_element(s.begin(), s.end());
    for (int i = 0; i <= N; i++) s[i] -= min_s;

    BIT<int> B(*max_element(s.begin(), s.end()) + 1);
    int64_t k = 0;
    for (int i = 0; i <= N; i++) {
        k += B.sum(s[i]);
        B.add(s[i], 1);
    }

    if (k >= ((int64_t)N*(N+1)/2 + 1)/2) return true;
    else return false;
}

int main() {
    cin >> N;
    a.resize(N);
    for (int i = 0; i < N; i++) cin >> a[i];

    int left = *min_element(a.begin(), a.end()) - 1;
    int right = *max_element(a.begin(), a.end()) + 1;

    while (right - left > 1) {
        int mid = (left + right) / 2;
        if (f(mid)) left = mid;
        else right = mid;
    }

    cout << left << endl;
}

File: problems/chapter06/mshibatatt/007.py
# https://atcoder.jp/contests/arc101/tasks/arc101_b
from itertools import accumulate
'''
median of m_lr where m_lr is median of a[l:r] (l <= r)
<=> max x s.t. n(m_lr < x) < (N(N+1)/2)//2)+1

<=> max x s.t. count of [l,r] < (N(N+1)/2)//2)+1
    s.t. n(a_i < x) > n(a_j >= x) where all a in [l, r]

<=> max x s.t. count of [l, r] < (N(N+1)/2)//2)+1  
    s.t. S_q_r - S_q_l-1 > 0 where S_q_i = sum_1^i q_i,
        where q_i = 1 if a_i < x else -1 

<=> max x s.t. count of [l, r] < (N(N+1)/2)//2)+1
    s.t. l <= r and  S_q_r > S_q_l-1,
        where S_q_i = sum_1^i q_i,
        where q_i = 1 if a_i < x else -1 

The condition is True if x = min(a)
The condition is False if x = max(a)
'''
class Bit():
    def __init__(self, n) -> None:
        self.size = n
        self.tree = [0] * (n + 1)
    
    def sum(self, i):
        s = 0
        while i > 0:
            s += self.tree[i]
            i -= i & -i
        return s
    
    def add(self, i, x):
        while i <= self.size:
            self.tree[i] += x
            i += i & -i
# 使用例
# bit = Bit(10)     # 要素数を与えてインスタンス化
# bit.add(2, 10)    # a2に10を加える
# print(bit.sum(3)) # a1～a3の合計を返す => 10

def main():
    N = int(input())
    a = list(map(int, input().split()))

    left = 1
    right = max(a) + 1
    threshold = (N*(N+1)/2)//2 + 1

    while right - left > 1:
        mid = left + (right - left)//2
        a_x = [0]+[1 if i < mid else -1 for i in a]
        qusai_inv = 0
        bit = Bit(2*N+2)
        
        for i, e in enumerate(accumulate(a_x)):
            bit.add(e+N+1, 1)
            qusai_inv += bit.sum(e+N)
            
        if qusai_inv < threshold:
            left = mid
        else:
            right = mid

    print(left)

if __name__=='__main__':
    main()